src/app/qgsprojectproperties.cpp

```diff
@@ -1439,12 +1442,35 @@ void QgsProjectProperties::apply()
 
   if ( grpWMSExt->isChecked() )
   {
-    QgsProject::instance()->writeEntry( QStringLiteral( "WMSExtent" ), QStringLiteral( "/" ),
-                                        QStringList()
-                                        << mWMSExtMinX->text()
-                                        << mWMSExtMinY->text()
-                                        << mWMSExtMaxX->text()
-                                        << mWMSExtMaxY->text() );
+    // Copy/paste from qgsextentwidget.cpp, so it is the same as the display
+    int decimals = 4;
+    switch ( mCrs.mapUnits() )
+    {
+      case QgsUnitTypes::DistanceDegrees:
+      case QgsUnitTypes::DistanceUnknownUnit:
+        decimals = 9;
+        break;
+      case QgsUnitTypes::DistanceMeters:
+      case QgsUnitTypes::DistanceKilometers:
+      case QgsUnitTypes::DistanceFeet:
+      case QgsUnitTypes::DistanceNauticalMiles:
+      case QgsUnitTypes::DistanceYards:
+      case QgsUnitTypes::DistanceMiles:
+      case QgsUnitTypes::DistanceCentimeters:
+      case QgsUnitTypes::DistanceMillimeters:
+        decimals = 4;
+        break;
+    }
+    // End of copy/paste
+    QgsRectangle wmsExtent = mAdvertisedExtentServer->outputExtent();
+    QgsProject::instance()->writeEntry(
+      QStringLiteral( "WMSExtent" ), QStringLiteral( "/" ),
+      QStringList()
+      << QString::number( wmsExtent.xMinimum(), 'f', decimals )
+      << QString::number( wmsExtent.yMinimum(), 'f', decimals )
+      << QString::number( wmsExtent.xMaximum(), 'f', decimals )
+      << QString::number( wmsExtent.yMaximum(), 'f', decimals )
```

Comment:
Could we not keep the maximum of decimal (or even more to be sure to avoid precision loss) whatever the map unit is ?

Reply:
As mentioned in the comment, it's a copy/paste from the QgsExtentWidget itself.
https://github.com/qgis/QGIS/blob/master/src/gui/qgsextentwidget.cpp#L227..L254

So I would argue that it's better for the end user perspective that we keep the same values between backend and display.
The user can write more precise extent manually. But then, when we would close the dialog, reopen the project, the precision would be lost anyway in the QLineEdit. The user might be confused that its value from UI is not the one saved in the XML.

Reply:
> The user can write more precise extent manually. But then, when we would close the dialog, reopen the project, the precision would be lost anyway in the QLineEdit. The user might be confused that its value from UI is not the one saved in the XML.

That would be still the case with the actual version no? Even worse when we are in meters for instance

> The user might be confused that its value from UI is not the one saved in the XML.

The one saved in the XML would be the same with trailing zeros.

Looks like we choose to write with a precision of 9 number in [processing](https://github.com/qgis/QGIS/blob/master/src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp#L5829) for instance

Reply:
In QGIS Server, we used 6 as wgs84 precision to display extent in GetCapabilities document, and methods to floor and ceil values with precision:
* wgs84 precision https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsgetcapabilities.cpp#L1572
* floor with precision https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsgetcapabilities.cpp#L1577
* ceil with precision https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsgetcapabilities.cpp#L1579

For the other CRS, the precision is set to 6 for Geographic and 3 for others. https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsgetcapabilities.cpp#L1673 so 4 and 9 are higher.